### PR TITLE
feat(weave): Project level leaderboards

### PR DESCRIPTION
**Pull request by tssweeney**
Project Level Leaderboard

See https://github.com/wandb/weave/blob/4e49cd3f9c564633f3959a6c82c2dad8ce262f4d/docs/docs/reference/gen_notebooks/leaderboard_quickstart.md for a full walkthrough

Screenshot 2024-10-28 at 21 58 48
Screenshot 2024-10-28 at 21 58 38

Land procedure:
1. Land https://github.com/wandb/weave/pull/2826 first
2. Extract the python changes, land those
3. Deploy Trace server
4. land the rest of this (Frontend and notebook)
5. Deploy FE

### CODE REVIEW

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/LeaderboardPage/SimpleCodeLikeTextArea.tsx [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+  // Automatically adjust height based on content
+  const adjustHeight = () => {
+    const textarea = textareaRef.current;
+    if (!textarea || isManuallyResized) {
+      return;
+    }
+
+    textarea.style.height = 'auto';
+    const lineHeight = parseInt(
+      getComputedStyle(textarea).lineHeight || '20',
```
Comment (gtarpenning) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/LeaderboardPage/SimpleCodeLikeTextArea.tsx:65]: wow `getComputedStyle(textarea).lineHeight` is a string??

File: weave-js/src/components/PagePanelComponents/Home/Browse3/views/Leaderboard/LeaderboardGrid.tsx [hunk 2]
```diff
@@ -111,8 +121,23 @@ export const LeaderboardGrid: React.FC<LeaderboardGridProps> = ({
...
+          let isOp = false;
+          try {
+            isOp =
+              Object.values(
+                Object.values(
+                  Object.values(
+                    (params.row as RowData).modelGroup.datasetGroups
+                  )[0].scorerGroups
+                )[0].metricPathGroups
+              )[0][0].modelType === 'op';
```
Comment (gtarpenning) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/views/Leaderboard/LeaderboardGrid.tsx:133]: i think we should make helpers for these deeply nested leaf value getters

**Review by gtarpenning — COMMENTED** [review]
This is a beast PR, not sure how best to review it. Checked it out and played with the example. Everything worked, a a screenshot of possibly fixable things. 

In general though I think the leaderboard flow is _confusing_. Clearly powerful, and probably useful, there is just a ton of configuration and numbers and API calls required to make it work. We might want to consider a happy-simple case and add a couple helpers around that. If we could do something similar to the Compare evals page where we populate at least **something** by default that might make this less initially obtuse. 


- duplicate columns behave strangely, they aren't displayed (which is fine) but do affect the mui sorting order numbers
Screenshot 2024-10-29 at 11 30 45 AM

File: tests/trace/test_leaderboard.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+                scorer_name="test_scorer_name",
+                summary_metric_path="test_summary_metric_path",
+            )
+        ],
+    )
+
+    ref = weave.publish(spec)
+
+    # TODO: this is not working
+    # spec = ref.get()
```
[the remark this answers is not in the document]
Comment (tssweeney) [reply tests/trace/test_leaderboard.py:32]: because the "ref" edges get deserialized to objects...

**Review by gtarpenning — APPROVED** [review]

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/LeaderboardPage/LeaderboardPage.tsx [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+export const LeaderboardPageContent: React.FC<
+  LeaderboardPageProps & {
+    setName: (name: string) => void;
+    isEditing: boolean;
+    setIsEditing: (isEditing: boolean) => void;
+  }
+> = props => {
+  const {useRootObjectVersions} = useWFHooks();
+  const {entity, project} = props;
+  const leaderboardObjectVersion = useRootObjectVersions(entity, project, {
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/LeaderboardPage/LeaderboardPage.tsx:90]: This should be moved to the new hook

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/LeaderboardPage/LeaderboardListingPage.tsx [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+  entity: string;
+  project: string;
+}> = props => {
+  const history = useHistory();
+  const {peekingRouter} = useWeaveflowRouteContext();
+  const {useRootObjectVersions} = useWFHooks();
+
+  // TODO: Once `useCollectionObjects` lands from the online
+  // evals project, switch to that (much more type safe)
+  const leaderboardObjectVersions = useRootObjectVersions(
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/LeaderboardPage/LeaderboardListingPage.tsx:108]: this should be moved to the new hook

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/LeaderboardPage/LeaderboardConfigEditor.tsx [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+
+type EvaluationHelperObj = {
+  obj: TraceObjSchema;
+  ref: string;
+  name: string;
+  versionIndex: number;
+  digest: string;
+};
+
+const useEvaluationObjects = (
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/LeaderboardPage/LeaderboardConfigEditor.tsx:346]: Extract to new file and use new hook

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/LeaderboardPage/LeaderboardPage.tsx [hunk 7]
```diff
@@ -0,0 +1,460 @@
...
+  return (
+    <LeaderboardPageContentInner
+      {...props}
+      leaderboardVal={leaderboardVal}
+      setIsEditing={props.setIsEditing}
+    />
+  );
+};
+
+const useUpdateLeaderboard = (
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/LeaderboardPage/LeaderboardPage.tsx:136]: extract to new hook

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/LeaderboardPage/SimpleCodeLikeTextArea.tsx (hunk 1 above)
Comment (tssweeney) [reply weave-js/src/components/PagePanelComponents/Home/Browse3/pages/LeaderboardPage/SimpleCodeLikeTextArea.tsx:65]: Screenshot 2024-10-31 at 12 47 49
 Indeed it is!

File: weave-js/src/components/PagePanelComponents/Home/Browse3/views/Leaderboard/LeaderboardGrid.tsx (hunk 2 above)
Comment (tssweeney) [reply weave-js/src/components/PagePanelComponents/Home/Browse3/views/Leaderboard/LeaderboardGrid.tsx:133]: Ack - fixed in next commit

**Pull request merged**